[REF] pos*: use bus.mixin

vlst-odoo opened the pull request:
pos*: point_of_sale, pos_adyen, pos_online_payment,
	  pos_restaurant, pos_self_order, pos_self_order_adyen

- create the `pos.bus.mixin` class;
	- it insures that modules that inherit this class have an
	  `access_token`; ( similar to `portal.mixin` )
	- implements the `_notify` method that uses the `access_token`
	  and the notification name concatenated as a channel;
	   --> as such, listeners can differentiate between senders.
- adapt `pos.config`, `pos.session`, `pos.preparation.display` to
  inherit from `pos.bus.mixin`;

- replace all bus usage in the frontend with `bus.subscribe`;
    ( the PosBus service has been removed )
  	- this has the benefit of solving the bug from task: 3557587

Task 3508655

https://github.com/odoo/enterprise/pull/47986




---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/136621).

vlst-odoo commented:
Hello @odoo/rd-security , In this pr we implement the suggestion from [here](https://github.com/odoo/odoo/pull/136467#issuecomment-1735422794). What do you think?

vlst-odoo commented on addons/point_of_sale/models/bus_mixin.py:
```diff
@@ -0,0 +1,21 @@
+
+    access_token = fields.Char('Security Token', copy=False)
+
+    def _bus_sendone(self, notification: str, message: dict):
+        self._bus_sendmany([(notification, message)])
+
+    def _bus_sendmany(self, notifications: list):
+        if not self.access_token:
+            # we use a `write` to force the cache clearing otherwise `return self.access_token` will return False
+            self.sudo().write({'access_token': str(uuid.uuid4())})
+        updated_notifications = [(self.access_token, notification, {"data": message, "channel": self.access_token}) for notification, message in notifications]
+        self.env['bus.bus']._sendmany(updated_notifications)
```
i agree, I kept `sendmany` just so it is the same as the method from `bus`. I will change it then

vlst-odoo replied:
it's done :+1:

vlst-odoo commented on addons/point_of_sale/models/bus_mixin.py:
```diff
@@ -0,0 +1,21 @@
+        if not self.access_token:
+            # we use a `write` to force the cache clearing otherwise `return self.access_token` will return False
+            self.sudo().write({'access_token': str(uuid.uuid4())})
+        updated_notifications = [(self.access_token, notification, {"data": message, "channel": self.access_token}) for notification, message in notifications]
```
you mean that the channel name is just the access token, as opposed to being `f"{model_name}-{access_token}"`?
I couldn't find a reason why i would also put the model name in the channel name. When we listen on the frontend we end up doing something like `bus.addChannel(this.pos.config.access_token)` so it becomes pretty clear that your message will be coming from the `pos.config`.

vlst-odoo replied:
about adding the channel name in the payload:
When using `bus_service` we don't know what channel a certain notification came from. We kept running into problems when 2 pos sessions were opened in 2 tabs. The shared worker handling websockets would send notifications to both tabs.

vlst-odoo replied:
i agree, i've now changed such that there is no more `private_bus_service` and we simply namespace the notifications according to their intended channels

vlst-odoo commented on addons/point_of_sale/static/src/app/store/private_bus_service.js:
```diff
@@ -0,0 +1,32 @@
+    },
+};
+
+registry.category("services").add("private_bus", privateBusService);
```
The main use for this service is to be able to listen just on notifications from your tab.

vlst-odoo replied:
we can indeed add more functionality to this service, but it's whole existence came from this very simple need of not mixing tabs. I figured that we could add more functionality as we come to need it in the future

vlst-odoo replied:
The service is no more

vlst-odoo commented on addons/point_of_sale/models/bus_mixin.py:
```diff
@@ -0,0 +1,21 @@
+
+
+class BusMixin(models.AbstractModel):
+    _name = "bus.mixin"
```
ok, we can namespace it, but wouldn't it make sense for this to be a general purpose model anyways?

vlst-odoo replied:
ok, it's renamed to `pos.bus.mixin`

vlst-odoo commented:
Hello @xmo-odoo , I addressed the review comments in a new commit. Do you agree with the changes? 
Also, do you think that the name `access_token` on the `pos.bus.mixin` is adequate? Would it pose problems if with the `pos.order` for ex, which inherits from both `portal.mixin` and `pos.bus.mixin` ?

vlst-odoo commented:
Hello @odoo/rd-security , Is this pr ok?

vlst-odoo commented:
Hello, @odoo/rd-security , Could you please take a look at this pr?

vlst-odoo commented:
Hello @odoo/rd-security , Is this pr ok in terms of security? All that happens is that we created a unified way of securely accessing the bus, following @xmo-odoo 's suggestion

vlst-odoo commented on addons/point_of_sale/models/pos_bus_mixin.py:
```diff
@@ -0,0 +1,15 @@
+    _name = "pos.bus.mixin"
+    _description = "Bus Mixin"
+
+    access_token = fields.Char('Security Token', readonly=True, default=lambda self: str(uuid.uuid4()))
```
Yes, you are right @xmo-odoo , this makes sense. The portal mixin does that with the `_portal_ensure_token` method and at first i wanted to do the same thing, but i wanted to prevent the case where someone tries to access the access_token field which has not already been computed. Is there a solution to this problem? Something similar to how we would write a getter in js?

vlst-odoo commented on addons/point_of_sale/models/pos_bus_mixin.py:
```diff
@@ -0,0 +1,15 @@
+
+    access_token = fields.Char('Security Token', readonly=True, default=lambda self: str(uuid.uuid4()))
+
+    def _notify(self, *notifications, private=True) -> None:
+        if isinstance(notifications[0], str):
```
i added an example of how to call this method. This would be the type hint: `*notifications: Union[str, Tuple[str, str]]`, but this does not reflect that in the case that you want to pass just one message,  `notifications` would have to contain exactly 2 strings...  
Also, assuming that the api is clear enough, isn't it the job of the caller to check that the call is correct? 
In any case: what do you think about this? 
```py
    def _notify(self, *notifications, private=True) -> None:
        """ Send a notification to the bus.
        ex: one notification: `self._notify('STATUS', {'status': 'closed'})`
            multiple notifications: `self._notify(('STATUS', {'status': 'closed'}), ('TABLE_ORDER_COUNT', {'count': 2}))`
        """
        if isinstance(notifications[0], str):
            if len(notifications) != 2:
                raise ValueError("If you want to send a single notification, you must provide a name: str and a message: any")
            notifications = [notifications]
        self.env['bus.bus']._sendmany((self.access_token, f"{self.access_token}-{name}" if private else name, message) for name, message in notifications)
```

vlst-odoo commented on addons/pos_self_order/models/pos_config.py:
```diff
@@ -401,17 +384,11 @@ def _compute_self_ordering_url(self):
-
-            return current_session_id.action_pos_session_closing_control()
+        self._notify('STATUS', {'status': 'closed'})
+        return self.current_session_id.action_pos_session_closing_control()
```
shouldn't i add `self.ensure_one()` in fact? You are always supposed to call `_notify` on exactly one record

vlst-odoo commented on addons/point_of_sale/static/src/app/store/pos_store.js:
```diff
@@ -86,6 +86,8 @@ export class PosStore extends Reactive {
     mainScreen = { name: null, component: null };
 
     static serviceDependencies = [
+        "orm",
```
check if orm is really needed

caburj reviewed (approved):
robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.
